web/src/app/(main)/games/[slug]/goals/_components/GoalCodeDialog.tsx

```diff
@@ -0,0 +1,233 @@
+            const response = await fetch('/api/goals/upload/replace', {
+                method: 'POST',
+                headers: {
+                    'Content-Type': 'application/json',
+                },
+                body: JSON.stringify({
+                    slug: slug,
+                    goals: parsedGoals,
+                }),
+            });
+
+            if (!response.ok) {
+                const error = await response.text();
+                throw new Error(`Failed to update goals: ${error}`);
+            }
+
+            // Refresh the goals list
+            mutateGoals();
+            notifyMessage('Goals updated successfully!');
+            close();
+        } catch (err) {
+            notifyMessage(
+                `Error saving goals: ${err instanceof Error ? err.message : 'Unknown error'}`,
+            );
+        } finally {
+            setIsLoading(false);
+        }
+    };
+
+    const handleCancel = () => {
+        setJsonValue(exportGoalsToJson());
+        setError(null);
+        setIsFullscreen(false);
+        close();
+    };
+
+    const toggleFullscreen = () => {
+        setIsFullscreen(!isFullscreen);
+    };
+
+    return (
+        <Dialog
+            open={isOpen}
+            onClose={handleCancel}
+            maxWidth={isFullscreen ? false : 'lg'}
+            fullWidth={!isFullscreen}
+            fullScreen={isFullscreen}
+        >
+            <DialogTitle>
+                <Box
+                    sx={{
+                        display: 'flex',
+                        justifyContent: 'space-between',
+                        alignItems: 'center',
+                    }}
+                >
+                    Edit Goals as JSON
+                    <IconButton onClick={toggleFullscreen} size="small">
+                        {isFullscreen ? (
+                            <FullscreenExitIcon />
+                        ) : (
+                            <FullscreenIcon />
+                        )}
+                    </IconButton>
+                </Box>
+            </DialogTitle>
+            <DialogContent>
+                <Box sx={{ mb: 2 }}>
+                    <Typography variant="body2" color="text.secondary">
+                        Edit your goals directly in JSON format.
+                    </Typography>
+                </Box>
+
+                <JsonEditor
+                    value={jsonValue}
+                    onChange={handleJsonChange}
+                    error={error}
+                    height={isFullscreen ? 'calc(100vh - 200px)' : 400}
+                    placeholder="Enter goals in JSON format..."
+                />
+
+                <Box
+                    sx={{
+                        mt: 2,
+                        display: 'flex',
+                        gap: 2,
+                        justifyContent: 'flex-end',
+                    }}
+                >
+                    <Button onClick={handleCancel} disabled={isLoading}>
```

Comment:
Stylistically, we usually have the cancel button use the error styling and the submit button use the success styling, especially on dialogs.